_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,256 @@
+- wharf-web [#13](https://github.com/iver-wharf/wharf-web/issues/13)
+
+## Explanation
+
+<details><summary>wharf-api</summary>
+
+The POST `/build/{buildid}/artifact` endpoint handles inserting artifacts.
+If there are TRX (XML) files, it also parses them to create an array of `TestResult` and
+one `TestResultSummary` per file.
+
+The summaries get inserted into the database table `test_result_summary`.
+The results get inserted into the database table `test_result`.
+
+![Database structure](https://github.com/iver-wharf/rfcs/blob/33e14dceb44755ce5e0c66b257bbffe2bb3568d7/_assets/wharf-db-graph.png)
+
+Pseudocode-like, without error handling
+```go
+// database_models.go
+// modified
+type Build struct {
+    BuildID     uint         `gorm:"primaryKey" json:"buildId"`
+    StatusID    BuildStatus  `gorm:"not null" json:"statusId"`
+    ProjectID   uint         `gorm:"not null;index:build_idx_project_id" json:"projectId"`
+    Project     *Project     `gorm:"foreignKey:ProjectID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    ScheduledOn *time.Time   `gorm:"nullable;default:NULL" json:"scheduledOn" format:"date-time"`
+    StartedOn   *time.Time   `gorm:"nullable;default:NULL" json:"startedOn" format:"date-time"`
+    CompletedOn *time.Time   `gorm:"nullable;default:NULL" json:"finishedOn" format:"date-time"`
+    GitBranch   string       `gorm:"size:300;default:'';not null" json:"gitBranch"`
+    Environment null.String  `gorm:"nullable;size:40" json:"environment" swaggertype:"string"`
+    Stage       string       `gorm:"size:40;default:'';not null" json:"stage"`
+    Params      []BuildParam `gorm:"foreignKey:BuildID" json:"params"`
+    IsInvalid   bool         `gorm:"not null;default:false" json:"isInvalid"` 
+    // added 
+    TestResultSummaryCount uint `gorm:"not null" json:"testResultSummaryCount"`
+}
+// new
+type TestResultSummary struct {
+    ArtifactID  uint	  `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
+    Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
+    Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    RunCount    uint	  `gorm:"not null" json:"runCount"`
+    SkipCount   uint	  `gorm:"not null" json:"skipCount"`
+    FailCount   uint	  `gorm:"not null" json:"failCount"`
+    PassCount   uint	  `gorm:"not null" json:"passCount"`
+}
+// new
+type TestResult struct {
+    ArtifactID  uint 	  `gorm:"not null;index:testresult_idx_artifact_id" json:"artifactId"`
+    Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    Name        string	  `gorm:"not null;" json:"name"`
+    Ran         string 	  `gorm:"not null;" json:"ran"`
+    Passed      string 	  `gorm:"not null;" json:"passed"`
```

Comment:
How does these properties, "Ran" & "Passed", work? They're both strings. What do they represent?

Reply:
Typo, meant to be bools.

Reply:
Suggest then to instead have an enum for it, something like "state" or "status" with the enum values of "skipped", "passed", & "failed"